python/cuml/preprocessing/TargetEncoder.py

```diff
@@ -0,0 +1,213 @@
+
+        agg_each_fold = train.groupby(cols).agg({self.y_col:
+                                                 ['count',
+                                                  'sum']}).reset_index()
+        agg_each_fold.columns = cols + ['count_y', 'sum_y']
+
+        agg_all = agg_each_fold.groupby(x_cols).agg({'count_y': 'sum',
+                                                     'sum_y': 'sum'})\
+                                               .reset_index()
+        cols = x_cols
+        agg_all.columns = cols + ['count_y_all', 'sum_y_all']
+
+        agg_each_fold = agg_each_fold.merge(agg_all, on=x_cols, how='left')
+        agg_each_fold['count_y_all'] = agg_each_fold['count_y_all'] - \
+            agg_each_fold['count_y']
+        agg_each_fold['sum_y_all'] = agg_each_fold['sum_y_all'] - \
+            agg_each_fold['sum_y']
+        agg_each_fold[self.out_col] = (agg_each_fold['sum_y_all'] +
+                                       self.smooth*self.mean) / \
+                                      (agg_each_fold['count_y_all'] +
+                                       self.smooth)
+        agg_each_fold = agg_each_fold.drop(['count_y_all',
+                                            'count_y',
+                                            'sum_y_all',
+                                            'sum_y'], axis=1)
+
+        agg_all[self.out_col] = (agg_all['sum_y_all'] + self.smooth *
+                                 self.mean) / \
+            (agg_all['count_y_all'] + self.smooth)
+        agg_all = agg_all.drop(['count_y_all', 'sum_y_all'], axis=1)
+        self.agg_all = agg_all
+
+        cols = [self.fold_col]+x_cols
+        train = train.merge(agg_each_fold, on=cols, how='left')
+        del agg_each_fold
+        return self._get_return_value(train), train
+
+    def _check_is_fitted(self):
+        if not self._fitted or self.train is None:
+            msg = ("This LabelEncoder instance is not fitted yet. Call 'fit' "
+                   "with appropriate arguments before using this estimator.")
+            raise NotFittedError(msg)
+
+    def _is_train_df(self, df):
+        if len(df) != len(self.train):
+            return False
+        for col in df.columns:
+            if col not in self.train.columns:
+                raise ValueError(f"Input column {col} "
+                                 "is not in train data.")
+            if not (df[col] == self.train[col]).all():
+                return False
+        return True
+
+    def _get_return_value(self, df):
+        df[self.out_col] = df[self.out_col].nans_to_nulls()
+        df[self.out_col] = df[self.out_col].fillna(self.mean)
+        df = df.sort_values(self.id_col)
+        res = df[self.out_col].values.copy()
+        return res
+
+    def _to_frame(self, x):
+        if isinstance(x, gd.DataFrame):
+            df = x.copy()
+        elif isinstance(x, gd.Series):
+            df = x.to_frame()
+        elif isinstance(x, cp.ndarray):
```

Comment:
would be good to support numpy and pandas inputs too as we do elsewhere...
Is there already an equivalent conversion utility in `input_utils`?

Reply:
Is it the best way here to adopt `cumlArray`?

Reply:
Ok, `cumlArray` doesn't support string columns right now. I'll do it manually.